keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1606 @@
+[Design Details](#design-details)) in its current form implements this first
+phase – the KEP will evolve and be supplemented with future phases getting
+implemented.
+
+![design](./design.svg)
+
+In the current design QoS-class resources are designed to be opaque to the CRI
+client in the sense that the container runtime takes care of configuration and
+control of the resources and the classes within.
+
+### Phase 1
+
+The goal is to enable a bare minimum for users to leverage QoS-class resources
+and start experimenting with them in Kubernetes:
+
+- extend the CRI protocol to allow QoS-class resource assignment and updates to
+  be communicated from kubelet to the runtime
+- extend the CRI protocol to allow runtime to communicate available QoS-class
+  resources (the types of resources and the classes within) to kubelet
+- implement pod annotations as an initial user interface
+- introduce a feature gate for enabling QoS-class resource support in kubelet
+
+### Future work
+
+This section sheds light on the end goal of this work in order to better
+evaluate this KEP in a broader context. What a fully working solution would
+consists of and what the (next) steps to accomplish that would be. These topics
+are currently listed as "future work" in [Goals](#goals).
+
+In practice, the future work mostly consists of changes to the Kubernetes API
+and control plane components.
+
+#### Pod Spec
+
+This future step will replace pod annotations with proper user interface in the
+Kubernetes API, i.e. PodSpec. Below, one possible option is presented.
+
+Introduce a new field (e.g. class) into ResourceRequirements of Container.
+
+```diff
+// ResourceRequirements describes the compute resource requirements.
+type ResourceRequirements struct {
+     // Limits describes the maximum amount of compute resources allowed.
+     Limits ResourceList `json:"limits,omitempty"
+     // Requests describes the minimum amount of compute resources required.
+     Requests ResourceList `json:"requests,omitempty"
++    // Classes specifies the QoS-class resources that the container should be assigned
++    Classes map[ClassResourceName]string
+}
+
++// ClassResourceName is the name of a QoS-class resource.
++type ClassResourceName string
+```
+
+Also, we add a `Resources` field to the `PodSpec`. We will re-use the existing
+`ResourceRequirements` type but Limits and Requests must be left empty. Classes
+may be set and they represent the Pod-level assignment of QoS-class resources,
+comparable to the PodClassResources message in PodSandboxConfig in the CRI API.
+
+```diff
+ type PodSpec struct {
+@@ -224,4 +224,8 @@ type PodSpec struct {
+     // Default to false.
+     // +optional
+     SetHostnameAsFQDN *bool `json:"setHostnameAsFQDN,omitempty" protobuf:"varint,35,opt,name=setHostnameAsFQDN"`
++    // Pod-level resources. Currently, requests and limits are not allowed
++    // to be specified for pods.
++    // +optional
++    Resources ResourceRequirements
```

Comment:
Can you share the reasoning on why the Resource model got adopted here?

What about a more exlicit API, i.e. following the runtimeClass API approach?

Reply:
@fabiand can you elaborate a bit here about your runtimeClass question?

Reply:
> @fabiand can you elaborate a bit here about your runtimeClass question?

This was a bit unclear to me, too. IF you mean just a dedicated "top-level" field (like `PodSpec.QosClassResources`) we thought about that. However, in `Container` we put it under `Resources` and wanted to align with that plus https://github.com/kubernetes/enhancements/pull/1592 which would bring `Resources` field in PodSpec level, too. We could have s dedicated field on the "top-level" but then we should do the same thing in `Container`, too. Whichever seems more logical. Was this answering your question?

In any case, this is currently "scheduled" to the next implementation phase so there is time to think about this.

Reply:
@fabiand any input?